test/extended/jenkins/jenkins_plugin.go

```diff
@@ -0,0 +1,160 @@
+	defer g.GinkgoRecover()
+	var oc = exutil.NewCLI("jenkins-plugin", exutil.KubeConfigPath())
+	var hostPort string
+
+	g.BeforeEach(func() {
+
+		g.By("set up policy for jenkins jobs")
+		err := oc.Run("policy").Args("add-role-to-user", "edit", "system:serviceaccount:"+oc.Namespace()+":default").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("kick off the build for the jenkins ephermeral and application templates")
+		jenkinsEphemeralPath := exutil.FixturePath("fixtures", "jenkins-ephemeral-template.json")
+		err = exutil.NewApp(jenkinsEphemeralPath, oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		jenkinsApplicationPath := exutil.FixturePath("fixtures", "application-template.json")
+		err = exutil.NewApp(jenkinsApplicationPath, oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("tag the app image stream")
+		err = exutil.TagImageStream("nodejs-010-centos7", "openshift/nodejs-010-centos7:latest", oc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("waiting for jenkins deployment")
+		err = exutil.WaitForADeployment(oc.KubeREST().ReplicationControllers(oc.Namespace()), "jenkins", exutil.CheckDeploymentCompletedFn, exutil.CheckDeploymentFailedFn)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("get ip and port for jenkins service")
+		serviceIP, err := oc.Run("get").Args("svc", "jenkins", "--config", exutil.KubeConfigPath()).Template("{{.spec.clusterIP}}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		port, err := oc.Run("get").Args("svc", "jenkins", "--config", exutil.KubeConfigPath()).Template("{{ $x := index .spec.ports 0}}{{$x.port}}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		hostPort = fmt.Sprintf("%s:%s", serviceIP, port)
+
+		g.By("wait for jenkins to come up")
+		err = waitForJenkinsActivity(fmt.Sprintf("http://%s", hostPort), "", 200)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+	})
+
+	g.Context("jenkins-plugin test context  ", func() {
+
+		g.It("jenkins-plugin test case execution", func() {
+
+			g.By("create jenkins job config xml file, convert to bytes for http post")
+			data := jenkinsJobBytes("testjob-plugin.xml", oc.Namespace())
+
+			g.By("make http request to create job")
+			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/createItem?name=test-plugin-job", hostPort), "POST", bytes.NewBuffer(data), 200)
+
+			g.By("make http request to kick off build")
+			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/job/test-plugin-job/build?delay=0sec", hostPort), "POST", nil, 201)
+
+			// the build and deployment is by far the most time consuming portion of the test jenkins job;
+			// we leverage some of the openshift utilities for waiting for the deployment before we poll
+			// jenkins for the sucessful job completion
+			g.By("waiting for frontend deployment as sign that the build has finished")
+			err := exutil.WaitForADeployment(oc.KubeREST().ReplicationControllers(oc.Namespace()), "frontend", exutil.CheckDeploymentCompletedFn, exutil.CheckDeploymentFailedFn)
```

Comment:
That doesn't strike me as sound reasoning... These are tests, everything
should be expected to potentially fail...

Ben Parees | OpenShift
On Dec 15, 2015 9:49 PM, "Steve Kuznetsov" <notifications@github.com> wrote:

> In test/extended/jenkins/jenkins_plugin.go
> <https://github.com/openshift/origin/pull/6329#discussion_r47731264>:
>
> > +		g.It("jenkins-plugin test case execution", func() {
> > +
> > +			g.By("create jenkins job config xml file, convert to bytes for http post")
> > +			data := jenkinsJobBytes("testjob-plugin.xml", oc.Namespace())
> > +
> > +			g.By("make http request to create job")
> > +			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/createItem?name=test-plugin-job", hostPort), "POST", bytes.NewBuffer(data), 200)
> > +
> > +			g.By("make http request to kick off build")
> > +			immediateInteractionWithJenkins(fmt.Sprintf("http://%s/job/test-plugin-job/build?delay=0sec", hostPort), "POST", nil, 201)
> > +
> > +			// the build and deployment is by far the most time consuming portion of the test jenkins job;
> > +			// we leverage some of the openshift utilities for waiting for the deployment before we poll
> > +			// jenkins for the sucessful job completion
> > +			g.By("waiting for frontend deployment as sign that the build has finished")
> > +			err := exutil.WaitForADeployment(oc.KubeREST().ReplicationControllers(oc.Namespace()), "frontend", exutil.CheckDeploymentCompletedFn, exutil.CheckDeploymentFailedFn)
>
> I believe the WaitFor methods were never used for something that was
> expected to fail so the timeouts on them were those imposed by Ginkgo on
> the rest itself.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6329/files#r47731264>.
>